### PR TITLE
Add Sources Module

### PR DESCRIPTION
**Pull request by lispandfound**
Adds the inaugural sources module to source_modelling repo.

### CODE REVIEW

File: wiki/Sources.md [hunk 1]
```diff
@@ -0,0 +1,213 @@
+# Source Geometry 
+
+A number of packages need to deal with fault geometry including:
```
Comment (AndrewRidden-Harper) [inline wiki/Sources.md:3]: Should this be "A number of packages **needed**..." ?

File: wiki/Sources.md [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+    return sp.optimize.minimize(
+        lambda x: source.fault_coordinates_to_wgs_depth_coordinates(x)[2],
+        np.array([1/2, 1/2]),
+        bounds=(0, 1)
+    )
+```
+
+## Answering Geometry Questions with the Sources Module
+
+Below is a kind of cookbook demonstrating how to use the new sources module to answer source geometry questions.
```
Comment (AndrewRidden-Harper) [inline wiki/Sources.md:136]: I think this will be a very useful section

**Review by AndrewRidden-Harper — APPROVED** [review]
Nice work! I didn't explicitly check any maths but didn't see any obvious problems when I glanced over it. Let me know if a more detailed check of the maths would be useful

File: wiki/Sources.md (hunk 1 above)
Comment (lispandfound) [reply wiki/Sources.md:3]: Yes it should be.

Comment (lispandfound) [reply wiki/Sources.md:3]: actually no it not should be.

File: source_modelling/sources.py [hunk 3]
```diff
@@ -0,0 +1,714 @@
...
+    """
+
+    def fault_coordinate_distance(fault_coordinates: np.ndarray) -> float:
+        source_a_global_coordinates = (
+            source_a.fault_coordinates_to_wgs_depth_coordinates(fault_coordinates[:2])
+        )
+        source_b_global_coordinates = (
+            source_b.fault_coordinates_to_wgs_depth_coordinates(fault_coordinates[2:])
+        )
+        return coordinates.distance_between_wgs_depth_coordinates(
```
Comment (lispandfound) [inline source_modelling/sources.py:699]: This is a function that either needs to be in qcore or I can copy it into this repo. I think a distance function between coordinates (like `ll_dist` but vectorised, more accurate, etc) is probably welcome in qcore?

File: wiki/Sources.md [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+
+## Sources Used in Ground Motion Simulation
+
+While we have five types of sources (per [Source Modelling for GMSim](https://wiki.canterbury.ac.nz/display/QuakeCore/Source+Modelling+for+GMSim)), there are essentially only three source geometries we work with:
+
+1. **Point Source Geometry**: This is a 0-dimensional geometry consisting of a single point. The `qcore.sources` module uses the `Point` class to model the source geometry for a point.
+2. **Plane Geometry**: This a 2-dimensional source geometry consisting of a single plane. The extents of the geometry are its corners. The `qcore.sources` module uses the `Plane` class to model single plane geometry.
+3. **Multi-Planar Geometry**: This is a 2-dimensional source geometry consisting of a number of connected planes. The extents of the geometry are the corners of the end fault planes. The `qcore.sources` module uses the `Fault` class to model multi-planar geometry.
+
+Type-I fault are an instance of the the first geometry, type-2 faults
```
Comment (sungeunbae) [inline wiki/Sources.md:90]: Type-1  (sorry, nitpicking!!)

File: source_modelling/sources.py (hunk 3 above)
Comment (joelridden) [reply source_modelling/sources.py:713]: I agree this could be welcomed into qcore if it's going to be used in multiple repos

**Review by joelridden — APPROVED** [review]

**Review by AndrewRidden-Harper — APPROVED** [review]

File: source_modelling/sources.py [hunk 5]
```diff
@@ -610,16 +675,18 @@ def wgs_depth_coordinates_to_fault_coordinates(
 
         """
         # the right edges as a cumulative proportion of the fault length (e.g. [0.1, ..., 0.8])
-        right_edges = self.lengths.cumsum() / self.length
-        right_edges = np.append(right_edges, 1)
+        left_edges = self.lengths.cumsum() / self.length
```
Comment (joelridden) [inline source_modelling/sources.py:678]: comment should probably change to left edge now

Comment (lispandfound) [reply source_modelling/sources.py:678]: ye

**Review by AndrewRidden-Harper — APPROVED** [review]
nice updates

**Review by sungeunbae — APPROVED** [review]

**Review by joelridden — APPROVED** [review]

**Pull request merged**